File: packages/shared/src/vehicleHealth.js
const DEFAULT_MONTHLY_MILES = 900;

export const VEHICLE_HEALTH_COMPONENTS = [
  {
    id: 'oil_change',
    label: 'Oil',
    intervalMiles: 5000,
    intervalDays: 180,
    costLow: 70,
    costHigh: 140,
  },
  {
    id: 'tire_rotation',
    label: 'Rotation',
    intervalMiles: 6000,
    intervalDays: 240,
    costLow: 25,
    costHigh: 60,
  },
  {
    id: 'tire_replacement',
    label: 'Tires',
    intervalMiles: 50000,
    intervalDays: 1460,
    costLow: 600,
    costHigh: 1400,
  },
  {
    id: 'brake_service',
    label: 'Brakes',
    intervalMiles: 40000,
    intervalDays: 1095,
    costLow: 300,
    costHigh: 900,
  },
  {
    id: 'battery_replacement',
    label: 'Battery',
    intervalMiles: 45000,
    intervalDays: 1460,
    costLow: 160,
    costHigh: 320,
  },
  {
    id: 'wiper_replacement',
    label: 'Wipers',
    intervalMiles: null,
    intervalDays: 365,
    costLow: 20,
    costHigh: 60,
  },
];

function parseMileage(value) {
  const parsed = Number.parseInt(String(value ?? '').replace(/[^0-9]/g, ''), 10);
  return Number.isFinite(parsed) ? parsed : 0;
}

function daysBetween(start, end) {
  const ms = end.getTime() - start.getTime();
  return Math.max(0, Math.round(ms / (1000 * 60 * 60 * 24)));
}

function addDays(date, days) {
  const copy = new Date(date.getTime());
  copy.setDate(copy.getDate() + days);
  return copy;
}

function clamp(value, min, max) {
  return Math.min(max, Math.max(min, value));
}

function normalizeText(...values) {
  return values
    .filter(Boolean)
    .map(value => String(value).toLowerCase())
    .join(' ');
}

export function inferHealthComponentIds(entry) {
  const text = normalizeText(
    entry?.serviceType,
    entry?.title,
    entry?.description,
    entry?.notes
  );

  const ids = [];

  if (/oil|filter change|lubrication/.test(text)) {
    ids.push('oil_change');
  }
  if (/tire rotation|rotate tires|rotate tire|rotated tires|rotated tire/.test(text)) {
    ids.push('tire_rotation');
  }
  if (
    /(replace|replacement|install|mounted).*(tire|tires)|(tire|tires).*(replace|replacement|install)/.test(
      text
    )
  ) {
    ids.push('tire_replacement');
  }
  if (/brake|brake pad|rotor|caliper/.test(text)) {
    ids.push('brake_service');
  }
  if (/battery/.test(text)) {
    ids.push('battery_replacement');
  }
  if (/wiper|windshield blade|washer blade/.test(text)) {
    ids.push('wiper_replacement');
  }

  return Array.from(new Set(ids));
}

function getMileageCadence(vehicle, maintenanceEntries, now) {
  const currentMileage = parseMileage(vehicle?.mileage);
  const points = maintenanceEntries
    .map(entry => {
      const mileage = parseMileage(entry?.mileage);
      const date = entry?.date ? new Date(entry.date) : null;
      if (!mileage || !date || Number.isNaN(date.getTime())) {
        return null;
      }
      return { mileage, date };
    })
    .filter(Boolean)
    .sort((a, b) => a.date.getTime() - b.date.getTime());

  if (currentMileage > 0) {
    points.push({ mileage: currentMileage, date: now });
  }

  if (points.length >= 2) {
    const first = points[0];
    const last = points[points.length - 1];
    const mileageDelta = Math.max(0, last.mileage - first.mileage);
    const dayDelta = Math.max(30, daysBetween(first.date, last.date));
    const monthly = Math.round((mileageDelta / dayDelta) * 30);
    if (monthly > 0) {
      return clamp(monthly, 200, 2500);
    }
  }

  const purchaseDate = vehicle?.purchaseDate ? new Date(vehicle.purchaseDate) : null;
  if (
    purchaseDate &&
    !Number.isNaN(purchaseDate.getTime()) &&
    currentMileage > 0 &&
    purchaseDate < now
  ) {
    const ownershipDays = Math.max(30, daysBetween(purchaseDate, now));
    const monthly = Math.round((currentMileage / ownershipDays) * 30);
    return clamp(monthly, 200, 2500);
  }

  return DEFAULT_MONTHLY_MILES;
}

function buildAnchor(componentId, vehicle, maintenanceEntries) {
  const matchingEntry = maintenanceEntries
    .filter(entry => inferHealthComponentIds(entry).includes(componentId))
    .sort((a, b) => {
      const aDate = new Date(a?.date || 0).getTime();
      const bDate = new Date(b?.date || 0).getTime();
      return bDate - aDate;
    })[0];

  if (matchingEntry) {
    return {
      source: 'record',
      recordId: matchingEntry.id,
      date: matchingEntry.date ? new Date(matchingEntry.date) : null,
      mileage: parseMileage(matchingEntry.mileage) || null,
      hasDate: Boolean(matchingEntry.date),
      hasMileage: parseMileage(matchingEntry.mileage) > 0,
    };
  }

  const purchaseDate = vehicle?.purchaseDate ? new Date(vehicle.purchaseDate) : null;
  if (purchaseDate && !Number.isNaN(purchaseDate.getTime())) {
    return {
      source: 'default',
      date: purchaseDate,
      mileage: 0,
      hasDate: true,
      hasMileage: false,
    };
  }

  return {
    source: 'default',
    date: null,
    mileage: null,
    hasDate: false,
    hasMileage: false,
  };
}

function toConfidenceBand(score) {
  if (score >= 0.75) return 'high';
  if (score >= 0.5) return 'medium';
  return 'low';
}

function getStatusFromPercent(percent) {
  if (percent <= 0) return 'overdue';
  if (percent <= 0.15) return 'service_soon';
  if (percent <= 0.35) return 'watch';
  return 'good';
}

function buildExplanation(component, anchor, status, milesPerMonth) {
  const base = [];
  if (anchor.source === 'record') {
    base.push('Based on your last recorded service entry.');
  } else {
    base.push('Using baseline interval assumptions because service history is limited.');
  }

  base.push(`Forecast assumes about ${milesPerMonth.toLocaleString()} miles per month.`);

  if (status !== 'good') {
    base.push('Keep mileage and service records current to improve estimate accuracy.');
  }

  return base;
}

export function computeVehicleHealthSnapshot(vehicle, maintenanceEntries = [], options = {}) {
  const now = options.now instanceof Date ? options.now : new Date();
  const currentMileage = parseMileage(vehicle?.mileage);
  const milesPerMonth = getMileageCadence(vehicle, maintenanceEntries, now);
  const milesPerDay = Math.max(1, Math.round(milesPerMonth / 30));

  const components = VEHICLE_HEALTH_COMPONENTS.map(component => {
    const anchor = buildAnchor(component.id, vehicle, maintenanceEntries);
    const elapsedMiles =
      anchor.mileage != null ? Math.max(0, currentMileage - anchor.mileage) : null;
    const elapsedDays =
      anchor.date != null ? daysBetween(anchor.date, now) : null;

    let remainingMiles = null;
    let remainingDays = null;
    let dueMileage = null;
    let dueDate = null;
    const percentCandidates = [];

    if (component.intervalMiles) {
      if (elapsedMiles != null) {
        remainingMiles = component.intervalMiles - elapsedMiles;
      } else if (currentMileage > 0) {
        const consumed = currentMileage % component.intervalMiles;
        remainingMiles = component.intervalMiles - consumed;
      }

      if (remainingMiles != null) {
        dueMileage = currentMileage + remainingMiles;
        percentCandidates.push(remainingMiles / component.intervalMiles);
      }
    }

    if (component.intervalDays) {
      if (elapsedDays != null) {
        remainingDays = component.intervalDays - elapsedDays;
      }

      if (remainingDays != null) {
        dueDate = addDays(now, remainingDays);
        percentCandidates.push(remainingDays / component.intervalDays);
      } else if (remainingMiles != null) {
        const convertedDays = Math.round(remainingMiles / milesPerDay);
        remainingDays = convertedDays;
        dueDate = addDays(now, convertedDays);
      }
    }

    const remainingLifePercent =
      percentCandidates.length > 0
        ? clamp(Math.min(...percentCandidates), -0.5, 1)
        : null;

    let confidenceScore = 0.35;
    if (anchor.source === 'record' && anchor.hasDate && anchor.hasMileage) {
      confidenceScore = 0.88;
    } else if (anchor.source === 'record' && (anchor.hasDate || anchor.hasMileage)) {
      confidenceScore = 0.68;
    } else if (currentMileage > 0) {
      confidenceScore = 0.4;
    }

    if (milesPerMonth !== DEFAULT_MONTHLY_MILES) {
      confidenceScore = clamp(confidenceScore + 0.05, 0, 0.95);
    }

    const status = getStatusFromPercent(remainingLifePercent ?? 0.2);

    return {
      componentId: component.id,
      label: component.label,
      status,
      confidenceScore,
      confidenceBand: toConfidenceBand(confidenceScore),
      remainingLifePercent,
      remainingMiles,
      remainingDays,
      estimatedDueMileage: dueMileage,
      estimatedDueDate: dueDate ? dueDate.toISOString().slice(0, 10) : null,
      estimatedCostLow: component.costLow,
      estimatedCostHigh: component.costHigh,
      anchorSource: anchor.source,
      anchorRecordId: anchor.recordId,
      explanation: buildExplanation(component, anchor, status, milesPerMonth),
    };
  });

  const weightedPercent = components.reduce((sum, component) => {
    const percent = component.remainingLifePercent ?? 0.4;
    return sum + clamp(percent, 0, 1);
  }, 0);

  let overallHealthScore = Math.round((weightedPercent / components.length) * 100);
  const overdueCount = components.filter(component => component.status === 'overdue').length;
  const soonCount = components.filter(
    component => component.status === 'service_soon'
  ).length;

  overallHealthScore -= overdueCount * 18;
  overallHealthScore -= soonCount * 8;
  overallHealthScore = clamp(overallHealthScore, 12, 99);

  const overallConfidenceScore =
    components.reduce((sum, component) => sum + component.confidenceScore, 0) /
    components.length;

  const rankedByUrgency = [...components].sort((a, b) => {
    const aDays = a.remainingDays ?? Number.POSITIVE_INFINITY;
    const bDays = b.remainingDays ?? Number.POSITIVE_INFINITY;
    if (aDays !== bDays) return aDays - bDays;
    const aMiles = a.remainingMiles ?? Number.POSITIVE_INFINITY;
    const bMiles = b.remainingMiles ?? Number.POSITIVE_INFINITY;
    return aMiles - bMiles;
  });

  const spendInWindow = months => {
    const horizonDays = months * 30;
    const horizonMiles = milesPerMonth * months;
    return rankedByUrgency.reduce(
      (totals, component) => {
        const dueInWindow =
          (typeof component.remainingDays === 'number' &&
            component.remainingDays <= horizonDays) ||
          (typeof component.remainingMiles === 'number' &&
            component.remainingMiles <= horizonMiles);

        if (!dueInWindow) {
          return totals;
        }

        return {
          low: totals.low + (component.estimatedCostLow || 0),
          high: totals.high + (component.estimatedCostHigh || 0),
        };
      },
      { low: 0, high: 0 }
    );
  };

  const spend90d = spendInWindow(3);
  const spend12m = spendInWindow(12);
  const spend36m = spendInWindow(36);

  const lowConfidenceCount = components.filter(
    component => component.confidenceBand === 'low'
  ).length;
  const missingServiceHistory = maintenanceEntries.length === 0;

  const accuracyTip = missingServiceHistory
    ? 'Add your recent oil, tire, brake, battery, and wiper service history to unlock higher-confidence forecasts.'
    : lowConfidenceCount > 0
      ? 'Keep mileage and service entries current so remaining-life estimates stay accurate.'
      : 'Your recent service history is helping keep these forecasts accurate.';

  return {
    vin: vehicle?.vin || '',
    generatedAt: now.toISOString(),
    overallHealthScore,
    overallConfidenceScore,
    overallConfidenceBand: toConfidenceBand(overallConfidenceScore),
    nextLikelyService: rankedByUrgency[0]?.label || null,
    estimatedMilesPerMonth: milesPerMonth,
    estimatedSpend90dLow: spend90d.low,
    estimatedSpend90dHigh: spend90d.high,
    estimatedSpend12mLow: spend12m.low,
    estimatedSpend12mHigh: spend12m.high,
    estimatedSpend36mLow: spend36m.low,
    estimatedSpend36mHigh: spend36m.high,
    components: rankedByUrgency,
    accuracyTip,
    missingServiceHistory,
    lowConfidenceCount,
  };
}

File: packages/mobile/lib/models/vehicle_health.dart
import 'dart:math' as math;

import 'maintenance.dart';
import 'vehicle.dart';

class VehicleHealthComponent {
  final String label;
  final String status;
  final String confidence;
  final double? remainingPercent;
  final int? remainingMiles;
  final int? remainingDays;
  final int? estimatedDueMileage;
  final DateTime? estimatedDueDate;
  final int costLow;
  final int costHigh;

  VehicleHealthComponent({
    required this.label,
    required this.status,
    required this.confidence,
    required this.remainingPercent,
    this.remainingMiles,
    this.remainingDays,
    this.estimatedDueMileage,
    this.estimatedDueDate,
    required this.costLow,
    required this.costHigh,
  });
}

class VehicleHealthSnapshot {
  final String vin;
  final String vehicleLabel;
  final int overallHealthScore;
  final String accuracyTip;
  final List<VehicleHealthComponent> components;
  final String? nextLikelyService;
  final int estimatedMilesPerMonth;
  final int estimatedSpend90dLow;
  final int estimatedSpend90dHigh;
  final int estimatedSpend12mLow;
  final int estimatedSpend12mHigh;
  final int estimatedSpend36mLow;
  final int estimatedSpend36mHigh;
  final bool missingServiceHistory;
  final int lowConfidenceCount;

  VehicleHealthSnapshot({
    required this.vin,
    required this.vehicleLabel,
    required this.overallHealthScore,
    required this.accuracyTip,
    required this.components,
    this.nextLikelyService,
    required this.estimatedMilesPerMonth,
    required this.estimatedSpend90dLow,
    required this.estimatedSpend90dHigh,
    required this.estimatedSpend12mLow,
    required this.estimatedSpend12mHigh,
    required this.estimatedSpend36mLow,
    required this.estimatedSpend36mHigh,
    required this.missingServiceHistory,
    required this.lowConfidenceCount,
  });
}

/// Component interval/cost table. Mirrors packages/shared/src/vehicleHealth.js's
/// VEHICLE_HEALTH_COMPONENTS exactly, so mobile and web compute the same
/// forecast from the same data. Wipers intentionally has no mileage interval,
/// matching the shared package.
class _HealthComponentSpec {
  final String id;
  final String label;
  final int? intervalMiles;
  final int intervalDays;
  final int costLow;
  final int costHigh;

  const _HealthComponentSpec({
    required this.id,
    required this.label,
    required this.intervalMiles,
    required this.intervalDays,
    required this.costLow,
    required this.costHigh,
  });
}

const List<_HealthComponentSpec> _kComponents = [
  _HealthComponentSpec(
    id: 'oil_change',
    label: 'Oil',
    intervalMiles: 5000,
    intervalDays: 180,
    costLow: 70,
    costHigh: 140,
  ),
  _HealthComponentSpec(
    id: 'tire_rotation',
    label: 'Rotation',
    intervalMiles: 6000,
    intervalDays: 240,
    costLow: 25,
    costHigh: 60,
  ),
  _HealthComponentSpec(
    id: 'tire_replacement',
    label: 'Tires',
    intervalMiles: 50000,
    intervalDays: 1460,
    costLow: 600,
    costHigh: 1400,
  ),
  _HealthComponentSpec(
    id: 'brake_service',
    label: 'Brakes',
    intervalMiles: 40000,
    intervalDays: 1095,
    costLow: 300,
    costHigh: 900,
  ),
  _HealthComponentSpec(
    id: 'battery_replacement',
    label: 'Battery',
    intervalMiles: 45000,
    intervalDays: 1460,
    costLow: 160,
    costHigh: 320,
  ),
  _HealthComponentSpec(
    id: 'wiper_replacement',
    label: 'Wipers',
    intervalMiles: null,
    intervalDays: 365,
    costLow: 20,
    costHigh: 60,
  ),
];

class _Anchor {
  final String source; // 'record' or 'default'
  final DateTime? date;
  final int? mileage;
  final bool hasDate;
  final bool hasMileage;

  const _Anchor({
    required this.source,
    this.date,
    this.mileage,
    required this.hasDate,
    required this.hasMileage,
  });
}

class VehicleHealthCalculator {
  static const int _defaultMonthlyMiles = 900;

  static int _parseMileage(dynamic value) {
    final parsed = int.tryParse(
      (value?.toString() ?? '').replaceAll(RegExp(r'[^0-9]'), ''),
    );
    return parsed ?? 0;
  }

  static int _daysBetween(DateTime start, DateTime end) {
    return math.max(0, end.difference(start).inDays);
  }

  static DateTime _addDays(DateTime date, int days) =>
      date.add(Duration(days: days));

  /// How many times a component recurs within a forecast window, given how
  /// soon it's next due and how often it repeats. Occurrence count is
  /// derived independently from the days-based and miles-based clocks
  /// (whichever the component uses), taking the larger of the two — a
  /// component due via either clock is priced for how often that clock
  /// actually recurs within the window, not just whether it's due at least
  /// once. Mirrors countOccurrencesInWindow in
  /// packages/shared/src/vehicleHealth.js.
  static int countOccurrencesInWindow({
    required int? remainingDays,
    required int? intervalDays,
    required int? remainingMiles,
    required int? intervalMiles,
    required int horizonDays,
    required int horizonMiles,
  }) {
    var occurrences = 0;

    if (remainingDays != null &&
        intervalDays != null &&
        intervalDays > 0 &&
        remainingDays <= horizonDays) {
      occurrences = math.max(
        occurrences,
        1 + ((horizonDays - remainingDays) / intervalDays).floor(),
      );
    }

    if (remainingMiles != null &&
        intervalMiles != null &&
        intervalMiles > 0 &&
        remainingMiles <= horizonMiles) {
      occurrences = math.max(
        occurrences,
        1 + ((horizonMiles - remainingMiles) / intervalMiles).floor(),
      );
    }

    return occurrences;
  }

  static double _clampD(double value, double min, double max) =>
      value < min ? min : (value > max ? max : value);

  /// Mirrors inferHealthComponentIds in packages/shared/src/vehicleHealth.js.
  static List<String> _inferComponentIds(Maintenance entry) {
    final text = '${entry.title} ${entry.notes}'.toLowerCase();
    final ids = <String>[];

    // Each of oil/brake/battery requires both a component term AND a
    // service-action term, the same way tire_replacement already does below
    // — a bare mention of the component ("brake noise inspection, no work
    // performed", "checked battery terminals, still good") must not reset
    // the forecast to "just serviced," or the app confidently reports a
    // safety-relevant component (brakes) as freshly done when nothing was
    // actually replaced.
    final hasServiceAction =
        text.contains('replace') ||
        text.contains('replacement') ||
        text.contains('replaced') ||
        text.contains('install') ||
        text.contains('installed') ||
        text.contains('change') ||
        text.contains('changed') ||
        text.contains('service') ||
        text.contains('serviced') ||
        text.contains('flush') ||
        text.contains('flushed');

    final hasOilTerm = text.contains('oil');
    if ((hasOilTerm && hasServiceAction) ||
        RegExp(r'filter change|lubrication').hasMatch(text)) {
      ids.add('oil_change');
    }
    if (RegExp(
      r'tire rotation|rotate tires|rotate tire|rotated tires|rotated tire',
    ).hasMatch(text)) {
      ids.add('tire_rotation');
    }
    final hasTireTerm = text.contains('tire') || text.contains('tires');
    final hasTireReplacementAction =
        text.contains('replace') ||
        text.contains('replacement') ||
        text.contains('install') ||
        text.contains('installed') ||
        text.contains('mounted');
    if (hasTireTerm && hasTireReplacementAction) {
      ids.add('tire_replacement');
    }
    final hasBrakeTerm =
        text.contains('brake') ||
        text.contains('brakes') ||
        text.contains('rotor') ||
        text.contains('rotors') ||
        text.contains('caliper') ||
        text.contains('calipers');
    final hasBrakeServiceAction =
        hasServiceAction || text.contains('pad') || text.contains('pads');
    if (hasBrakeTerm && hasBrakeServiceAction) {
      ids.add('brake_service');
    }
    final hasBatteryServiceAction = hasServiceAction || text.contains('new');
    if (text.contains('battery') && hasBatteryServiceAction) {
      ids.add('battery_replacement');
    }
    if (RegExp(r'wiper|windshield blade|washer blade').hasMatch(text)) {
      ids.add('wiper_replacement');
    }

    return ids.toSet().toList();
  }

  /// Mirrors getMileageCadence in packages/shared/src/vehicleHealth.js.
  static int _getMileageCadence(
    Vehicle vehicle,
    List<Maintenance> maintenance,
    DateTime now,
  ) {
    final currentMileage = vehicle.mileage;
    final points = <({int mileage, DateTime date})>[];

    for (final entry in maintenance) {
      final mileage = _parseMileage(entry.mileage);
      if (mileage <= 0) continue;
      points.add((mileage: mileage, date: entry.date));
    }
    points.sort((a, b) => a.date.compareTo(b.date));

    if (currentMileage > 0) {
      points.add((mileage: currentMileage, date: now));
    }

    if (points.length >= 2) {
      final first = points.first;
      final last = points.last;
      final mileageDelta = math.max(0, last.mileage - first.mileage);
      final dayDelta = math.max(30, _daysBetween(first.date, last.date));
      final monthly = ((mileageDelta / dayDelta) * 30).round();
      if (monthly > 0) {
        return _clampD(monthly.toDouble(), 200, 2500).round();
      }
    }

    final purchaseDateStr = vehicle.purchaseDate;
    if (purchaseDateStr != null && purchaseDateStr.isNotEmpty) {
      final purchaseDate = DateTime.tryParse(purchaseDateStr);
      if (purchaseDate != null &&
          currentMileage > 0 &&
          purchaseDate.isBefore(now)) {
        final ownershipDays = math.max(30, _daysBetween(purchaseDate, now));
        final monthly = ((currentMileage / ownershipDays) * 30).round();
        return _clampD(monthly.toDouble(), 200, 2500).round();
      }
    }

    return _defaultMonthlyMiles;
  }

  /// Mirrors buildAnchor in packages/shared/src/vehicleHealth.js, including
  /// the (0 vs null) mileage distinction: a matched record with no mileage
  /// leaves the mileage anchor unset (null), but the purchase-date fallback
  /// deliberately anchors mileage at 0 (assumes the component was serviced
  /// new at purchase), so both platforms treat an unserviced component the
  /// same way once a purchase date is known.
  static _Anchor _buildAnchor(
    String componentId,
    Vehicle vehicle,
    List<Maintenance> maintenance,
  ) {
    // Entries with a fabricated date (see Maintenance.hasKnownDate) sort as
    // if very old, mirroring the JS implementation's `new Date(entry.date ||
    // 0)` — otherwise a record with an unknown date would default to
    // DateTime.now() and wrongly win "most recent" against a genuinely
    // dated real record.
    final epoch = DateTime.fromMillisecondsSinceEpoch(0);
    final matching =
        maintenance
            .where((entry) => _inferComponentIds(entry).contains(componentId))
            .toList()
          ..sort((a, b) {
            final aDate = a.hasKnownDate ? a.date : epoch;
            final bDate = b.hasKnownDate ? b.date : epoch;
            return bDate.compareTo(aDate);
          });

    if (matching.isNotEmpty) {
      final entry = matching.first;
      final parsedMileage = _parseMileage(entry.mileage);
      return _Anchor(
        source: 'record',
        date: entry.hasKnownDate ? entry.date : null,
        mileage: parsedMileage > 0 ? parsedMileage : null,
        hasDate: entry.hasKnownDate,
        hasMileage: parsedMileage > 0,
      );
    }

    final purchaseDateStr = vehicle.purchaseDate;
    if (purchaseDateStr != null && purchaseDateStr.isNotEmpty) {
      final purchaseDate = DateTime.tryParse(purchaseDateStr);
      if (purchaseDate != null) {
        return _Anchor(
          source: 'default',
          date: purchaseDate,
          mileage: 0,
          hasDate: true,
          hasMileage: false,
        );
      }
    }

    return const _Anchor(
      source: 'default',
      date: null,
      mileage: null,
      hasDate: false,
      hasMileage: false,
    );
  }

  static String _confidenceBand(double score) {
    if (score >= 0.75) return 'high';
    if (score >= 0.5) return 'medium';
    return 'low';
  }

  static String _statusFromPercent(double? percent) {
    final p = percent ?? 0.2;
    if (p <= 0) return 'overdue';
    if (p <= 0.15) return 'service soon';
    if (p <= 0.35) return 'watch';
    return 'good';
  }

  static VehicleHealthSnapshot buildSnapshot(
    Vehicle vehicle,
    List<Maintenance> maintenance, {
    DateTime? now,
  }) {
    final currentTime = now ?? DateTime.now();
    final currentMileage = vehicle.mileage;
    final milesPerMonth = _getMileageCadence(vehicle, maintenance, currentTime);
    final milesPerDay = math.max(1, (milesPerMonth / 30).round());

    final components = <VehicleHealthComponent>[];

    for (final spec in _kComponents) {
      final anchor = _buildAnchor(spec.id, vehicle, maintenance);
      final elapsedMiles = anchor.mileage != null
          ? math.max(0, currentMileage - anchor.mileage!)
          : null;
      final elapsedDays = anchor.date != null
          ? _daysBetween(anchor.date!, currentTime)
          : null;

      int? remainingMiles;
      int? remainingDays;
      int? dueMileage;
      DateTime? dueDate;
      final percentCandidates = <double>[];

      if (spec.intervalMiles != null) {
        if (elapsedMiles != null) {
          remainingMiles = spec.intervalMiles! - elapsedMiles;
        } else if (currentMileage > 0) {
          final consumed = currentMileage % spec.intervalMiles!;
          remainingMiles = spec.intervalMiles! - consumed;
        }

        if (remainingMiles != null) {
          dueMileage = currentMileage + remainingMiles;
          percentCandidates.add(remainingMiles / spec.intervalMiles!);
        }
      }

      if (elapsedDays != null) {
        remainingDays = spec.intervalDays - elapsedDays;
      }

      if (remainingDays != null) {
        dueDate = _addDays(currentTime, remainingDays);
        percentCandidates.add(remainingDays / spec.intervalDays);
      } else if (remainingMiles != null) {
        final convertedDays = (remainingMiles / milesPerDay).round();
        remainingDays = convertedDays;
        dueDate = _addDays(currentTime, convertedDays);
      }

      final remainingPercent = percentCandidates.isNotEmpty
          ? _clampD(
              percentCandidates.reduce((a, b) => a < b ? a : b),
              -0.5,
              1.0,
            )
          : null;

      double confidenceScore = 0.35;
      if (anchor.source == 'record' && anchor.hasDate && anchor.hasMileage) {
        confidenceScore = 0.88;
      } else if (anchor.source == 'record' &&
          (anchor.hasDate || anchor.hasMileage)) {
        confidenceScore = 0.68;
      } else if (currentMileage > 0) {
        confidenceScore = 0.4;
      }
      if (milesPerMonth != _defaultMonthlyMiles) {
        confidenceScore = _clampD(confidenceScore + 0.05, 0, 0.95);
      }

      final status = _statusFromPercent(remainingPercent);

      components.add(
        VehicleHealthComponent(
          label: spec.label,
          status: status,
          confidence: _confidenceBand(confidenceScore),
          remainingPercent: remainingPercent,
          remainingMiles: remainingMiles,
          remainingDays: remainingDays,
          estimatedDueMileage: dueMileage,
          estimatedDueDate: dueDate,
          costLow: spec.costLow,
          costHigh: spec.costHigh,
        ),
      );
    }

    // Overdue severity is reflected once, by letting remainingPercent go
    // negative (down to its own -0.5 floor) in the average — not by also
    // subtracting a flat per-item penalty on top, which previously
    // double-counted overdue/soon items and bottomed every
    // heavily-neglected vehicle out at the same floor score regardless of
    // how overdue it actually was.
    final weightedPercent = components.fold<double>(
      0,
      (sum, component) =>
          sum + (component.remainingPercent ?? 0.4).clamp(-0.5, 1),
    );
    var overallHealthScore = ((weightedPercent / components.length) * 100)
        .round();
    overallHealthScore = overallHealthScore.clamp(0, 100);

    final rankedByUrgency = [...components]
      ..sort((a, b) {
        const infinity = 1 << 30;
        final aDays = a.remainingDays ?? infinity;
        final bDays = b.remainingDays ?? infinity;
        if (aDays != bDays) return aDays.compareTo(bDays);
        final aMiles = a.remainingMiles ?? infinity;
        final bMiles = b.remainingMiles ?? infinity;
        return aMiles.compareTo(bMiles);
      });

    final specByLabel = {for (final spec in _kComponents) spec.label: spec};

    // Recurring items (oil, rotation, wipers) recur multiple times within a
    // long window — a 36-month oil forecast should price in ~6 changes, not
    // 1.
    ({int low, int high}) spendInWindow(int months) {
      final horizonDays = months * 30;
      final horizonMiles = milesPerMonth * months;
      var low = 0;
      var high = 0;
      for (final component in rankedByUrgency) {
        final spec = specByLabel[component.label];
        final occurrences = countOccurrencesInWindow(
          remainingDays: component.remainingDays,
          intervalDays: spec?.intervalDays,
          remainingMiles: component.remainingMiles,
          intervalMiles: spec?.intervalMiles,
          horizonDays: horizonDays,
          horizonMiles: horizonMiles,
        );
        if (occurrences == 0) continue;
        low += component.costLow * occurrences;
        high += component.costHigh * occurrences;
      }
      return (low: low, high: high);
    }

    final spend90d = spendInWindow(3);
    final spend12m = spendInWindow(12);
    final spend36m = spendInWindow(36);

    final lowConfidenceCount = components
        .where((component) => component.confidence == 'low')
        .length;
    final missingServiceHistory = maintenance.isEmpty;

    final accuracyTip = missingServiceHistory
        ? 'Add your recent service entries to improve forecast accuracy.'
        : 'Keep mileage and service entries current so remaining-life estimates stay accurate.';

    return VehicleHealthSnapshot(
      vin: vehicle.vin,
      vehicleLabel: '${vehicle.year} ${vehicle.make} ${vehicle.model}',
      overallHealthScore: overallHealthScore,
      accuracyTip: accuracyTip,
      components: rankedByUrgency,
      nextLikelyService: rankedByUrgency.isNotEmpty
          ? rankedByUrgency.first.label
          : null,
      estimatedMilesPerMonth: milesPerMonth,
      estimatedSpend90dLow: spend90d.low,
      estimatedSpend90dHigh: spend90d.high,
      estimatedSpend12mLow: spend12m.low,
      estimatedSpend12mHigh: spend12m.high,
      estimatedSpend36mLow: spend36m.low,
      estimatedSpend36mHigh: spend36m.high,
      missingServiceHistory: missingServiceHistory,
      lowConfidenceCount: lowConfidenceCount,
    );
  }
}
